db/view.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/client/request"
+	"github.com/sourcenetwork/defradb/datastore"
+	"github.com/sourcenetwork/defradb/db/description"
+)
+
+func (db *db) addView(
+	ctx context.Context,
+	txn datastore.Txn,
+	inputQuery string,
+	sdl string,
+) ([]client.CollectionDefinition, error) {
+	// Wrap the given query as part of the GQL query object - this simplifies the syntax for users
+	// and ensures that we can't be given mutations.  In the future this line should disappear along
+	// with the all calls to the parser appart from `ParseSDL` when we implement the DQL stuff.
+	query := fmt.Sprintf(`query { %s }`, inputQuery)
+
+	newDefinitions, err := db.parser.ParseSDL(ctx, sdl)
+	if err != nil {
+		return nil, err
+	}
+
+	ast, err := db.parser.BuildRequestAST(query)
+	if err != nil {
+		return nil, err
+	}
+
+	req, errs := db.parser.Parse(ast)
+	if len(errs) > 0 {
+		return nil, errors.Join(errs...)
+	}
+
+	if len(req.Queries) == 0 || len(req.Queries[0].Selections) == 0 {
+		return nil, NewErrInvalidViewQuery(inputQuery)
+	}
+
+	baseQuery, ok := req.Queries[0].Selections[0].(*request.Select)
+	if !ok {
+		return nil, NewErrInvalidViewQuery(inputQuery)
+	}
+
+	for i, definition := range newDefinitions {
+		definition.Description.BaseQuery = baseQuery
+		newDefinitions[i] = definition
+	}
```

Comment:
+1

Reply:
lol, any reason why you see this as a suggestion and not a personal preference nitpick?

Reply:
For me it's the unnecessary allocation and that it makes it harder to read what's happening. It's just one extra line but you have to pause to think for a second what is happening instead of just seeing that you set something for every new definition.

Reply:
Fair enough :)  Will change.

- [ ] simplify loop

Reply:
> "you wrote shit, take this beauty instead"

:rofl: 

Generally, if I feel like something is just personal preference, and that I might be suggesting it just because I am more used to something, I'll mark it as a nitpick, and I actively try and limit the number of nitpicks I raise - for example IIRC in your PRs you seem to prefer setting the capacity with `make([]foo{}, 0, x)` and then using `append`, I find that a bit jarring to read and think of it as more error prone because it is more unusual, although there is nothing inherently wrong about it.

Todo's are anything I'd block a merge over (although can/should be argued against be the author if they disagree). Suggestions are everything else :)